### PR TITLE
Fleshed out Go's documentation about the director feature and added a director example.

### PR DESCRIPTION
**Pull request by michael-schaller**
Fixes issues #418.

### CODE REVIEW

**Comment by michael-schaller** [comment]
@ianlancetaylor 
Ian, it took ages to flesh out the documentation of the director feature.  I sincerely hope that you like the direction and detail of the new documentation.  I've used a very simple C++ example to explain the most important parts around the director feature and added the example to this commit as well.  I didn't update the existing 'callback' and 'extend' examples.  Please let me know what we should do with these (update or remove).

**Comment by michael-schaller** [comment]
So the Travis run failed with this error:
```
...
checking Examples/go/director
cp: cannot stat `director.go': No such file or directory
...
make: *** [check-go-examples] Error 2
```

Running this check on my relatively newly installed machine with Ubuntu 15.04 amd64 and SWIG tip I'm seeing no error:
```
$ make check-go-examples && echo SUCCESS || echo FAIL
...
SUCCESS
```

I wonder why the check-go-examples target failed on Travis on something as simple as a cp. Especially because this cp should copy example.go and director.go which should be in the same directory. director.go is included in the commit and so the log doesn't help me in any way to diagnose this issue.

Ian, does this example work for you? If yes, could you point me at the issue? If no, is there a way for me to troubleshoot this further on Travis?

**Comment by ojwb** [comment]
@michael-schaller The travis build runs with builddir != srcdir (it builds in subdirectory `build/build`) - since `example.go` is generated, it'll be in builddir, but director.go will be in srcdir.  I think that is your problem, and you are presumably building in srcdir locally, so it works for you.

**Comment by michael-schaller** [comment]
@ojwb 
Olly, thank you for pointing me in the right direction. That was indeed the issue.

**Comment by michael-schaller** [comment]
Finally everything is green. That took me unexpectedly long... :-/

**Comment by ianlancetaylor** [comment]
Will look at this when I'm back from vacation--thanks.

**Comment by michael-schaller** [comment]
Take your time, Ian. Wish you a nice remaining vacation...

**Comment by michael-schaller** [comment]
@ianlancetaylor Friendly ping.  I hope you had a nice vacation.

File: Doc/Manual/Go.html [hunk 1]
```diff
@@ -573,49 +583,547 @@
...
-C++ class with virtual methods.  This is complicated by the fact that
-C++ and Go define inheritance differently.  In Go, structs can inherit
-methods via anonymous field embedding.  However, when a method is
-called for an embedded struct, if that method calls any other methods,
-they are called for the embedded struct, not for the original type.
-Therefore, SWIG must use Go interfaces to represent C++ inheritance.
+SWIG's director feature permits a Go type to act as the subclass of a C++ class.
+This is complicated by the fact that C++ and Go define inheritance differently.
+SWIG normally represents the C++ class inheritance automatically in Go via
+interfaces but with a Go type representing a subclass of a C++ class quite some
```
Comment (ianlancetaylor) [inline Doc/Manual/Go.html]: s/quite//

File: Doc/Manual/Go.html [hunk 2]
```diff
@@ -573,49 +583,547 @@
...
+</p>
+
+
+<H4><a name="Go_director_ctor_dtor"></a>23.4.7.3 Constructor and destructor</H4>
+
+
+<p>
+SWIG creates an additional set of constructor and destructor functions once the
+director feature has been enabled for a C++ class.  <tt>NewDirectorClassName
+</tt> allows to override virtual methods on the new object instance and <tt>
```
Comment (ianlancetaylor) [inline Doc/Manual/Go.html]: Looksl ike the </tt> should move to the end of the previous line.

Comment (ianlancetaylor) [inline Doc/Manual/Go.html]: s/allows to override/allows overriding/

Comment (ianlancetaylor) [inline Doc/Manual/Go.html]: Looks like the trailing tt should move the start of the next line.

File: Doc/Manual/Go.html [hunk 3]
```diff
@@ -573,49 +583,547 @@
...
+methods all overriden protected virtual C++ methods will be public in Go.
 </p>
 
+<p>
+As an example see part of the <tt>FooBarGo</tt> class:
+</p>
+
+<div class="code">
+<pre>
+type overwritenMethodsOnFooBarAbs struct {
```
Comment (ianlancetaylor) [inline Doc/Manual/Go.html]: Seems like the type should be called overwrittenMethodsOnFooBarAbs -- that is, two 't's in "overwritten".

File: Doc/Manual/Go.html [hunk 4]
```diff
@@ -573,49 +583,547 @@
...
+</p>
+
+
+<H4><a name="Go_director_base_methods"></a>23.4.7.5 Call base methods</H4>
+
+
+<p>
+Often a virtual method will be overwritten to extend the original behavior of
+the method in the base class.  This is also the case for the <tt>FooBarCpp::Foo
+</tt> method of the example code:
```
Comment (ianlancetaylor) [inline Doc/Manual/Go.html]: The leading /tt should probably be at the end of the previous line.

File: Doc/Manual/Go.html (hunk 1 above)
Comment (michael-schaller) [inline Doc/Manual/Go.html]: done

File: Doc/Manual/Go.html (hunk 2 above)
Comment (michael-schaller) [inline Doc/Manual/Go.html]: done

File: Doc/Manual/Go.html (hunk 3 above)
Comment (michael-schaller) [inline Doc/Manual/Go.html]: done
```
$ find . -type f | xargs sed -i 's/overwritenMethodsOnFooBarAbs/overwrittenMethodsOnFooBarAbs/g'
```

File: Doc/Manual/Go.html (hunk 4 above)
Comment (michael-schaller) [inline Doc/Manual/Go.html]: done

**Comment by michael-schaller** [comment]
@ianlancetaylor PTAL

**Comment by ianlancetaylor** [comment]
LGTM.  Thanks for working on this.  Sorry for the very slow review.

**Pull request merged**